### PR TITLE
ESLint rule - no server modules on client side

### PR DESCRIPTION
**Pull request by LetItRock**
ESLint rule that doesn't allow to use the Node.js built-in modules in client-side code  - react components.
This feature is related to [the comment from @leerob](https://github.com/vercel/next.js/issues/27051#issuecomment-884139234) in #27051

@leerob fyi

The rule is extendable, devs can define their own forbidden modules using the options ex: `forbiddenImports: ['cron']`.

Some screenshots:

Screenshot 2021-08-07 at 23 58 32

Screenshot 2021-08-07 at 23 58 51

Screenshot 2021-08-08 at 00 01 51

Screenshot 2021-08-07 at 23 59 29

Screenshot 2021-08-07 at 23 59 51

Screenshot 2021-08-08 at 00 00 05

## Feature
- [x] ESLint rule
- [x] ESLint rule tests added
- [ ] Implements an existing feature request or RFC. Make sure the feature request has been accepted for implementation before opening a PR.
- [ ] Related issues linked using `fixes #number`
- [ ] Integration tests added
- [ ] Documentation added
- [ ] Telemetry added. In case of a feature if it's used or not.
- [ ] Errors have helpful link attached, see `contributing.md`

## Documentation / Examples

- [x] Make sure the linting passes

### CODE REVIEW

File: errors/no-server-modules-on-client-side.md [hunk 1]
```diff
@@ -0,0 +1,55 @@
+# Do Not Use Node.js Built in Modules in Client Side Code.
```
Comment (LetItRock) [inline errors/no-server-modules-on-client-side.md:1]: we can totally change this with your suggestions ;)

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| buildDuration | 14.7s | 14.9s | ⚠️ +164ms |
| buildDurationCached | 3.4s | 3.4s | -39ms |
| nodeModulesSize | 50.2 MB | 50.2 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.49 | 2.524 | ⚠️ +0.03 |
| / avg req/sec | 1004.1 | 990.51 | ⚠️ -13.59 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.407 | 1.4 | -0.01 |
| /error-in-render avg req/sec | 1777.38 | 1785.97 | +8.59 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.1 kB | 23.1 kB | ✓ |
| webpack-HASH.js gzip | 1.5 kB | 1.5 kB | ✓ |
| Overall change | 67 kB | 67 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _app-HASH.js gzip | 980 B | 980 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.52 kB | 2.52 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 904 B | 904 B | ✓ |
| image-HASH.js gzip | 4.12 kB | 4.12 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 319 B | 319 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 12.8 kB | 12.8 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _buildManifest.js gzip | 490 B | 490 B | ✓ |
| Overall change | 490 B | 490 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| buildDuration | 11.9s | 11.7s | -192ms |
| buildDurationCached | 4.6s | 4.5s | -83ms |
| nodeModulesSize | 50.2 MB | 50.2 MB | ✓ |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 2.565 | 2.574 | ⚠️ +0.01 |
| / avg req/sec | 974.74 | 971.2 | ⚠️ -3.54 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 1.445 | 1.451 | ⚠️ +0.01 |
| /error-in-render avg req/sec | 1730.35 | 1723.18 | ⚠️ -7.17 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| 17.HASH.js gzip | 185 B | 185 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14 kB | 14 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.6 kB | 10.6 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _app-HASH.js gzip | 965 B | 965 B | ✓ |
| _error-HASH.js gzip | 3.71 kB | 3.71 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 2.97 kB | 2.97 kB | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 2.94 kB | 2.94 kB | ✓ |
| withRouter-HASH.js gzip | 294 B | 294 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 17.7 kB | 17.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _buildManifest.js gzip | 498 B | 498 B | ✓ |
| Overall change | 498 B | 498 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| index.html gzip | 577 B | 577 B | ✓ |
| link.html gzip | 589 B | 589 B | ✓ |
| withRouter.html gzip | 571 B | 571 B | ✓ |
| Overall change | 1.74 kB | 1.74 kB | ✓ |

</details>

</details>
Commit: e046a6e61baa09ed79322a80f9ca4746932b5c41

**Comment by ijjk** [comment]
# Stats from current PR

<details>
<summary><strong>Default Build</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| buildDuration | 18s | 17.9s | -87ms |
| buildDurationCached | 4.1s | 4s | -78ms |
| nodeModulesSize | 50.2 MB | 50.2 MB | -3 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.193 | 3.193 | ✓ |
| / avg req/sec | 783.07 | 782.97 | ⚠️ -0.1 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.006 | 2.027 | ⚠️ +0.02 |
| /error-in-render avg req/sec | 1246.41 | 1233.64 | ⚠️ -12.77 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| 745.HASH.js gzip | 179 B | 179 B | ✓ |
| framework-HASH.js gzip | 42.2 kB | 42.2 kB | ✓ |
| main-HASH.js gzip | 23.1 kB | 23.1 kB | ✓ |
| webpack-HASH.js gzip | 1.5 kB | 1.5 kB | ✓ |
| Overall change | 67 kB | 67 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.1 kB | 31.1 kB | ✓ |
| Overall change | 31.1 kB | 31.1 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _app-HASH.js gzip | 980 B | 980 B | ✓ |
| _error-HASH.js gzip | 194 B | 194 B | ✓ |
| amp-HASH.js gzip | 312 B | 312 B | ✓ |
| css-HASH.js gzip | 329 B | 329 B | ✓ |
| dynamic-HASH.js gzip | 2.52 kB | 2.52 kB | ✓ |
| head-HASH.js gzip | 350 B | 350 B | ✓ |
| hooks-HASH.js gzip | 904 B | 904 B | ✓ |
| image-HASH.js gzip | 4.12 kB | 4.12 kB | ✓ |
| index-HASH.js gzip | 261 B | 261 B | ✓ |
| link-HASH.js gzip | 1.66 kB | 1.66 kB | ✓ |
| routerDirect..HASH.js gzip | 319 B | 319 B | ✓ |
| script-HASH.js gzip | 387 B | 387 B | ✓ |
| withRouter-HASH.js gzip | 320 B | 320 B | ✓ |
| bb14e60e810b..30f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 12.8 kB | 12.8 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _buildManifest.js gzip | 490 B | 490 B | ✓ |
| Overall change | 490 B | 490 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| index.html gzip | 531 B | 531 B | ✓ |
| link.html gzip | 543 B | 543 B | ✓ |
| withRouter.html gzip | 524 B | 524 B | ✓ |
| Overall change | 1.6 kB | 1.6 kB | ✓ |

</details>

</details>
<hr/>
<details>
<summary><strong>Webpack 4 Mode</strong> (Decrease detected ✓)</summary>

<br/>

<details>
<summary><strong>General</strong> Overall decrease ✓</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| buildDuration | 14.3s | 14.2s | -64ms |
| buildDurationCached | 5.6s | 5.5s | -172ms |
| nodeModulesSize | 50.2 MB | 50.2 MB | -3 B |

</details>

<details>
<summary><strong>Page Load Tests</strong> Overall decrease ⚠️</summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| / failed reqs | 0 | 0 | ✓ |
| / total time (seconds) | 3.119 | 3.193 | ⚠️ +0.07 |
| / avg req/sec | 801.57 | 783.07 | ⚠️ -18.5 |
| /error-in-render failed reqs | 0 | 0 | ✓ |
| /error-in-render total time (seconds) | 2.066 | 2.062 | 0 |
| /error-in-render avg req/sec | 1210.08 | 1212.64 | +2.56 |

</details>

<details>
<summary><strong>Client Bundles (main, webpack, commons)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| 17.HASH.js gzip | 185 B | 185 B | ✓ |
| 677f882d2ed8..HASH.js gzip | 14 kB | 14 kB | ✓ |
| framework.HASH.js gzip | 41.9 kB | 41.9 kB | ✓ |
| main-HASH.js gzip | 10.6 kB | 10.6 kB | ✓ |
| webpack-HASH.js gzip | 1.19 kB | 1.19 kB | ✓ |
| Overall change | 67.8 kB | 67.8 kB | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 31.3 kB | 31.3 kB | ✓ |
| Overall change | 31.3 kB | 31.3 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _app-HASH.js gzip | 965 B | 965 B | ✓ |
| _error-HASH.js gzip | 3.71 kB | 3.71 kB | ✓ |
| amp-HASH.js gzip | 552 B | 552 B | ✓ |
| css-HASH.js gzip | 333 B | 333 B | ✓ |
| dynamic-HASH.js gzip | 2.71 kB | 2.71 kB | ✓ |
| head-HASH.js gzip | 2.97 kB | 2.97 kB | ✓ |
| hooks-HASH.js gzip | 911 B | 911 B | ✓ |
| index-HASH.js gzip | 231 B | 231 B | ✓ |
| link-HASH.js gzip | 1.64 kB | 1.64 kB | ✓ |
| routerDirect..HASH.js gzip | 298 B | 298 B | ✓ |
| script-HASH.js gzip | 2.94 kB | 2.94 kB | ✓ |
| withRouter-HASH.js gzip | 294 B | 294 B | ✓ |
| e025d2764813..52f.css gzip | 125 B | 125 B | ✓ |
| Overall change | 17.7 kB | 17.7 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| _buildManifest.js gzip | 498 B | 498 B | ✓ |
| Overall change | 498 B | 498 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | LetItRock/next.js feature/eslint_rule_no_server_modules_on_client_side | Change |
| - | - | - | - |
| index.html gzip | 577 B | 577 B | ✓ |
| link.html gzip | 589 B | 589 B | ✓ |
| withRouter.html gzip | 571 B | 571 B | ✓ |
| Overall change | 1.74 kB | 1.74 kB | ✓ |

</details>

</details>
Commit: 4b0f7c4a438b5e2bb1b996ba504124ef79c160ab

**Comment by balazsorban44** [comment]
Hi @LetItRock, I honestly missed this PR before starting to work on https://github.com/vercel/next.js/pull/32410 😬.

I took a different approach though, only checking if those Node.js APIs are being used in one of the data fetching methods instead.

Are you still working on this PR? If so, then maybe I should close mine. Let me know

**Comment by balazsorban44** [comment]
Hi, I will close this PR for now, as there seems to be no activity on it, and there are merge conflicts as well. @LetItRock thanks for the work so far! If you wish to pick this up and get it up-to-date, feel free to! Otherwise, anyone else is welcome to continue. :pray: :green_heart:

**Pull request closed without merging**